ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
I'm not sure this is quite right. What if we are deliberately using a version locked method and the session version is different? Or what if we are using a version flexible method, but the session version is disabled via SSL_OP_NO_*?